lib/Ryu/Async.pm

```diff
@@ -439,7 +439,16 @@ sub udp_server {
         service  => $uri->port // 0,
         socktype => 'dgram'
     )->then(sub {
-        Future->done($server->write_handle->sockport)
+        Future->done($server->write_handle->sockport);
+    })->on_fail(sub {
```

Comment:
Should this be sequencing from `then()`, or should we be using the two-arg form of `then(\&on_done, \&on_fail)` here?

Reply:
indeed, at the moment this change doesn't really *do* anything :sweat_smile:

Reply:
However it works fine and exception can be catched in the script.

Reply:
No, unfortunately it *doesn't* work - the exception is triggered whenever the Future happens to be resolved, you can't rely on it being at the point you call this code.

Throwing an exception within `->on_fail` is almost always the wrong thing to do: it is quite likely to result in the event loop having to deal with the exception.

Instead, the second parameter to `->then` should be used if the Future failed and we want a better exception for it. That'd be available via `->port`, e.g. `try { my $srv = $ryu->udp_server(...); $srv->port->get } catch ($e) { warn $e->message }`.

Note that the existing `->port` Future should *already* be usable for this, it'd just have whatever error was left over from the bind call.